Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,329 @@
+  ext x
+  simp only [Function.comp_apply, Pi.zero_apply]
+  rw [h]
+  exact Set.mem_range_self x
+
+open LinearMap
+
+variable {R : Type _} [CommSemiring R]
+variable {M N P : Type _}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommMonoid P]
+  [Module R M] [Module R N] [Module R P]
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+lemma Exact.linearMap_ker_eq (hfg : Exact f g) :
+  ker g = range f := by
+  ext y
+  exact hfg y
+
+lemma LinearMap.exact_iff :
+  Exact f g ↔ LinearMap.ker g = LinearMap.range f := by
+  constructor
+  · intro h; apply Exact.linearMap_ker_eq; exact h
+  · intro h y; rw [← mem_ker, h]; rfl
+
+lemma Exact.linearMap_comp_eq_zero (h : Exact f g) : g.comp f = 0 := by
+  ext x
+  simp only [coe_comp, Function.comp_apply, zero_apply]
+  rw [h]
+  exact Set.mem_range_self x
+
+end Exact
+
+section TensorProduct
+
+open TensorProduct LinearMap
+
+section Semiring
+
+variable {R : Type _} [CommSemiring R]
+variable {M N P P' : Type _}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommGroup P] [AddCommMonoid P']
+  [Module R M] [Module R N] [Module R P] [Module R P']
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+variable {Q : Type _} [AddCommMonoid Q] [Module R Q]
+
+variable (g)
+
+lemma le_comap_range_lTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (lTensor Q g)).comap (TensorProduct.mk R Q P q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨q ⊗ₜ[R] n, rfl⟩
+
+lemma le_comap_range_rTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (rTensor Q g)).comap
+      ((TensorProduct.mk R P Q).flip q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨n ⊗ₜ[R] q, rfl⟩
+
+variable (Q) {g}
+
+theorem rTensor.Surjective (hg : Function.Surjective g) :
+  Function.Surjective (rTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | C0 => use 0; rw [map_zero]
+  | C1 p q =>
+      obtain ⟨n, rfl⟩ := hg p
+      use n ⊗ₜ[R] q
+      simp only [rTensor_tmul]
+  | Cp x y hx hy =>
+      obtain ⟨x, rfl⟩ := hx
+      obtain ⟨y, rfl⟩ := hy
+      use x + y
+      rw [map_add]
+
+theorem lTensor.Surjective (hg : Function.Surjective g) :
+  Function.Surjective (lTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | C0 => use 0; rw [map_zero]
+  | C1 q p =>
+      obtain ⟨n, rfl⟩ := hg p
+      use q ⊗ₜ[R] n
+      simp only [lTensor_tmul]
+  | Cp x y hx hy =>
+      obtain ⟨x, rfl⟩ := hx
+      obtain ⟨y, rfl⟩ := hy
+      use x + y
+      rw [map_add]
+
+end Semiring
+
+variable {R : Type _} [CommRing R]
+variable {M N P P' : Type _}
+  [AddCommGroup M] [AddCommGroup N] [AddCommGroup P]
+  [Module R M] [Module R N] [Module R P]
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+variable (Q : Type _) [AddCommGroup Q] [Module R Q]
+
+variable (hfg : Exact f g) (hg : Function.Surjective g)
+
+private noncomputable
+def rTensor.inverse_ofRightInverse {h : P → N} (hgh : Function.RightInverse h g) :
+  P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ (LinearMap.range (rTensor Q f)) := by
+  rw [exact_iff] at hfg
```

Comment:
A general style comment
```suggestion
def rTensor.inverse_ofRightInverse {h : P → N} (hgh : Function.RightInverse h g) :
    P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ (LinearMap.range (rTensor Q f)) := by
  rw [exact_iff] at hfg
```
and throughout

Reply:
OK, I had (apparently wrongly) understood that hypotheses had a 4 char indent, but the goal only 2. I'll fix that.

Reply:
Done.